### ISSUE TITLE
Looking for advice for GetChanges/ApplyChanges on complex graph

### ISSUE THREAD

**Issue by stevedelong** [tonysneed/trackable-entities#71]
I've got a complex object graph that I'm trying to use GetChanges/ApplyChanges to reduce the amount of data I need to send back to the server to save changes on the client, and I seem to be running into a problem with how it walks through my data to find changes.  I'm passing around a fairly hefty amount of data (this isn't a couple of rows or a simple tree or anything, so I understand it's not necessarily playing to all of EF's strengths), and that I believe is why I'm having this problem.

Specifically, I have a root object containing multiple parent > child > grandchild > greatgrandchild > yougettheidea hierarchies, all objects have parent pointers, and there's a fair amount of cross-linking as well.  On a very small data set I seem to be running into endless, or near-endless processing trying to get or apply changes (and am in danger of running into some truly scary deep graphs that I'll have to adjust WCF's max allowed depth in serialization to compensate for).

I've considered modifying Get/Apply to look for a new custom attribute that I would apply to appropriate navigation properties in my model, and if that attribute is found then it would not look for changes across that boundary (allowing me to limit the search depth because I know the other side of that boundary will be checked later), but obviously that adds a layer of maintenance I need to do to my model whenever it changes.

Short of breaking up the object graph into pieces to physically limit the change search (which I'd prefer not to do if at all possible), is there any guidance you can give me on this?

Thanks!

**Comment by tonysneed** [comment]
Hi Steve,

Do all the changes need to be persisted in a single transaction?  If not, you might want to consider batching them or performing them in parallel.

@azabluda, I'm curious about ideas you might have to parallelize ```GetChanges``` or ```ApplyChanges```?  I imagine you could take a large object graph and break it up into pieces that could be processed in parallel.

Let's say, however, that you were able to speed up getting and applying changes.  You have another performance problem when it comes time to persist the updates by calling ```DbContext.SaveChanges```.  That's because Entity Framework will execute an update for each and every entity that has been changed, resulting in an equal number of round trips to the database.  This can kill performance, especially if all the entities are included in the same transaction and the isolation level is high.

So you might want to look at some alternatives to ```DbContext.SaveChanges```, for example, ```DbContext.ExecuteStoreCommand``` and passing native SQL that is dynamically generated.  In that case, you would need a method that is similar to ```ApplyChanges``` that traverses the object graph to generate the SQL you need to send to the database.  In this case, a large number of updates would impact database performance through locking, but at least the number of round trips would be reduced.  So a big part of addressing the issue depends on your transactional requirements.

Cheers,
Tony

**Comment by azabluda** [comment]
The very first thing I would do is identify the performance bottleneck. From the discussion it isn't quite clean which method is the major evil doer:
* GetChanges (or perhaps Json serializer?)
* ApplyChanges
* DbContext.SaveChanges

Parallel processing is always nice of course, but it can at best shorten the execution time down by factor CpuCount. If we face some more fundamental defects (e.g. polynomial or even exponential computation complexity), then we should first revisit the algorithms rather than plunging into the fun world of parallel computing :)

I think we should first build a test scenario with lots of interrelated objects and see where exactly the performance goes down. Maybe @stevedelong could submit a unit test to us?

If Steve needs more control over the property visitation process, then he might consider implementing https://github.com/tonysneed/trackable-entities/blob/release-2.5.2/Source/TrackableEntities.Common/INavigationPropertyInspector.cs in his entities.

**Comment by stevedelong** [comment]
Another developer and I dug into this a bit more and have found out that the performance issue was on our end and we're now seeing that GetChanges is missing some values due to how our model is shaped (so if you'd prefer I can make a new issue about what I'm seeing now).

Here's a simple (and hopefully representative) model sample.

class Root
{
	ChangeTrackingCollection<DataSet> DataSets { get; set; }
}

class DataSet
{
	Root Root { get; set; }
	ChangeTrackingCollection<DisplayInfo> DisplayInfos { get; set; }
	ChangeTrackingCollection<LabelInfo> LabelInfos { get; set; }
	ChangeTrackingCollection<ValueType> ValueTypes { get; set; }
}

class DisplayInfo
{
	DataSet DataSet { get; set; }
	ChangeTrackingCollection<DisplayItem> DisplayItems { get; set; }
}

class DisplayItem
{
	DisplayInfo DisplayInfo { get; set; }
	ValueType ValueType { get; set; }
}

class LabelInfo
{
	DataSet DataSet { get; set; }
	ChangeTrackingCollection<LabelItem> LabelItems { get; set; }
}

class LabelItem
{
	LabelInfo LabelInfo { get; set; }
	ValueType ValueType { get; set; }
}

class ValueType
{
	DataSet DataSet { get; set; }
	ChangeTrackingCollection<Value> Values { get; set; }
}

class Value
{
	ValueType ValueType { get; set; }
}

If I have a simple case (one Root, with one DataSet, with one DisplayInfo, LabelInfo, ValueType, one DisplayItem and LabelItem for that ValueType), and add a new Value, then when I GetChanges it appears this is what happens:

1) Root > DataSet > DisplayInfo > DisplayItem > ValueType > LabelItem > LabelInfo > found no changes
2) Root > DataSet > ValueType is skipped, already visited, so Value is never found.

As @azabluda recommended I've looked at INavigationPropertyInspector and I'm going to give that a try next, but I wanted to follow up with the issue here as well so you knew what I was seeing now that I had a better understanding of the issue.

**Comment by tonysneed** [comment]
Go ahead and open a **new issue** for this one, so we can track the performance aspect separately.  To report the bug :bug:, it would be great if you could supply us with a **unit test**.  If you fork the latest develop branch, you'll see we are now using [xUnit.net](http://xunit.github.io/), which is the de facto standard and is well documented.

For guidelines on how to go about forking the repo and creating a pull request containing your unit test, check out the project guidelines for [contributing](http://tonysneed.github.io/trackable-entities/6-contributing.html).  Thanks in advance!

**Comment by stevedelong** [comment]
Sounds good -- I'll close this one now, and once I've got a good unit test to reproduce what we're seeing I'll open a new issue for it with unit tests.

Thanks for the help!

**Comment by harrisunderwork** [comment]
@stevedelong - Hi Steven, I had same issue on my code where the database is literally made of around 900 tables having same structure more or less same as yours. I made some simple changes in GetChanges as well as ApplyChanges(). We can discuss this in more detail if you like.

**Comment by stevedelong** [comment]
@harrisunderwork - I've come up with an implementation in my code that has things working as I want them to, but would be curious to see your changes in case they do a more robust job as I go forward.  Thanks for the offer, I'm curious to see what you've done.
